rust/src/utils.rs

```diff
@@ -304,6 +309,87 @@ pub(crate) fn translate_path_for_ostree_impl(path: &str) -> Option<String> {
+pub fn deduplicate_tmpfiles_entries(tmprootfs_dfd: i32) -> CxxResult<()> {
+    let tmprootfs_dfd = unsafe { ffiutil::ffi_dirfd(tmprootfs_dfd)? };
+    static TMPFILESD: &str = "usr/lib/tmpfiles.d";
+    static RPMOSTREE_TMPFILESD: &str = "usr/lib/rpm-ostree/tmpfiles.d";
+    static AUTOVAR_PATH: &str = "rpm-ostree-autovar.conf";
+
+    // read all rpm-ostree auto generated entries and save
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(RPMOSTREE_TMPFILESD)
+        .context("readdir {RPMOSTREE_TMPFILESD}")?;
+    let mut rpmostree_tmpfiles_entries: HashMap<String, String> =
+        save_tmpfile_entries(&tmpfiles_dir)?
+            .into_iter()
+            .map(|s| {
+                let entry = scan_tmpfile_entries(&s).unwrap();
+                (entry, s.to_string())
+            })
+            .collect::<HashMap<String, String>>();
+
+    // remove autovar.conf first, then read all system entries and save
+    let tmpfiles_dir = tmprootfs_dfd
+        .open_dir(TMPFILESD)
+        .context("readdir {TMPFILESD}")?;
+
+    if tmpfiles_dir.try_exists(AUTOVAR_PATH)? {
+        tmpfiles_dir.remove_file(AUTOVAR_PATH)?;
+    }
+    let system_tmpfiles_entries: HashSet<String> = save_tmpfile_entries(&tmpfiles_dir)?
+        .into_iter()
+        .map(|s| scan_tmpfile_entries(&s).unwrap())
+        .collect::<HashSet<String>>();
+
+    // remove duplicated entries in auto-generated tmpfiles.d,
+    // which are already in system tmpfiles
+    for key in system_tmpfiles_entries.into_iter() {
+        rpmostree_tmpfiles_entries.retain(|k, _value| k != &key);
+    }
+
+    {
+        // save the noduplicated entries
+        let mut entries = String::from("# This file was generated by rpm-ostree.\n");
+        for (_key, value) in rpmostree_tmpfiles_entries {
+            entries.push_str(&format!("{}\n", value));
+        }
+
+        let perms = Permissions::from_mode(0o644);
+        tmpfiles_dir.atomic_write_with_perms(&AUTOVAR_PATH, entries.as_bytes(), perms)?;
+    }
+    Ok(())
+}
+
+#[context("Save tmpfiles entries")]
+pub(crate) fn save_tmpfile_entries(tmpfiles_dir: &Dir) -> Result<Vec<String>> {
+    let mut entries: Vec<String> = Vec::new();
+    for entry in tmpfiles_dir.entries()? {
+        let entry = entry?;
+        let name = entry.file_name();
+        if !name.to_string_lossy().ends_with(".conf") {
+            continue;
+        }
+        let mut tmp_entries = tmpfiles_dir
+            .read_to_string(name)
+            .unwrap()
+            .lines()
+            .map(|s| s.to_string())
+            .filter(|s| !s.is_empty() && !s.starts_with('#'))
+            .collect::<Vec<String>>();
+
+        entries.append(&mut tmp_entries);
+    }
+    Ok(entries)
+}
+
+#[context("Scan tmpfiles entries and get path")]
+pub(crate) fn scan_tmpfile_entries(line: &str) -> Result<String> {
+    let parts: Vec<&str> = line.split_whitespace().collect();
+    let entry = parts.get(1).unwrap();
```

Comment:
On the first line - we also don't need to make an intermediate vector.  This can just be:

```
  line.split_whitespace().nth(1).ok_or_else(|| anyhow::anyhow!("Malformed tmpfiles.d entry {line}"))
```
or so?